src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala

```diff
@@ -0,0 +1,82 @@
+package com.wavesplatform.state2.reader
+
+import cats.implicits._
+import com.wavesplatform.state2._
+import scorex.account.{Account, Alias}
+import scorex.transaction.assets.exchange.ExchangeTransaction
+import scorex.transaction.lease.LeaseTransaction
+import scorex.transaction.{Transaction, TransactionParser}
+
+import scala.collection.JavaConverters.iterableAsScalaIterableConverter
+
+class StateReaderImpl(p: JavaMapStorage) extends StateReader {
+
+  override def transactionInfo(id: ByteArray): Option[(Int, Transaction)] = Option(p.transactions.get(id.arr)).map {
+    case (h, bytes) => (h, TransactionParser.parseBytes(bytes).get)
+  }
+
+  override def accountPortfolio(a: Account): Portfolio = {
+    Option(p.portfolios.get(a.bytes)).map { case (b, (i, o), as) => Portfolio(b, LeaseInfo(i, o), as.map { case (k, v) => EqByteArray(k) -> v }) }.orEmpty
+  }
+
+  override def assetInfo(id: ByteArray): Option[AssetInfo] = Option(p.assets.get(id.arr)).map {
+    case (is, amt) => AssetInfo(is, amt)
+  }
+
+  override def height: Int = p.getHeight
+
+  override def accountTransactionIds(a: Account): Seq[ByteArray] = {
+    Option(p.accountTransactionIds.get(a.bytes))
+      .map(_.toSeq)
+      .getOrElse(Seq.empty)
+      .map(EqByteArray)
+  }
+
+  override def nonEmptyAccounts: Seq[Account] =
+    p.portfolios
+      .keySet()
+      .asScala
+      .map(b => Account.fromBytes(b).right.get)
```

Comment:
I don't think I can to much about it.
Yes, in this particular case it's possible to flatten/whatever, but the code generally relies on data stored being correct. This includes deserialising txs,accs, other 'smart-constructed' entities from state.